network-area-diagram-viewer/src/main/resources/svg.js

```diff
@@ -0,0 +1,463 @@
+    var debugSvg = svg.getElementById(debugRotationId);
+    if (!debugSvg) {
+        // https://www.motiontricks.com/creating-dynamic-svg-elements-with-javascript/
+        var polyline = document.createElementNS(SVG_NAMESPACE, "polyline");
+        polyline.setAttribute("id", debugRotationId);
+        polyline.setAttribute("points", "0, 0 50, 0 40, 10 50, 0, 40, -10");
+        polyline.setAttribute("style", "fill:none;stroke:" + DEBUG_COLOR[debugRotationId] + ";stroke-width:2");
+        svg.appendChild(polyline);
+        console.log("created debug element [" + debugRotationId + "]");
+        debugSvg = polyline
+    }
+    return debugSvg;
+}
+
+function debugRotation(svg, cx, cy, angle, debugRotationId) {
+    debugSvg = getCreateDebugRotation(svg, debugRotationId);
+    if (debugSvg) {
+        translate(svg, debugSvg, {x: cx, y: cy});
+        var transforms = debugSvg.transform.baseVal;
+        if (transforms.length === 1) {
+            rotate = svg.createSVGTransform();
+            debugSvg.transform.baseVal.appendItem(rotate);
+        }
+        transforms.getItem(1).setRotate(angle, 0, 0);
+    }
+}
+
+function debugConnection(svg, edgeSvg, edgeId, p0, p1) {
+    var debugConnectionId = edgeId + "-debug";
+    var debugSvg = svg.getElementById(debugConnectionId);
+    if (!debugSvg) {
+        var polyline = document.createElementNS(SVG_NAMESPACE, "polyline");
+        polyline.setAttribute("id", debugConnectionId);
+        polyline.setAttribute("points", p0.x + ", " + p0.y + " " + p1.x + " " + p1.y);
+        polyline.setAttribute("style", "fill:none;stroke:" + DEBUG_COLOR["debug-connection"] + ";stroke-width:2");
+        svg.appendChild(polyline);
+        console.log("created debug element [" + debugConnectionId + "]");
+    } else {
+        var polyline = debugSvg;
+        polyline.points.clear();
+        var p = svg.createSVGPoint();
+        p.x = p0.x;
+        p.y = p0.y;
+        polyline.points.appendItem(p);
+        p = svg.createSVGPoint();
+        p.x = p1.x;
+        p.y = p1.y;
+        polyline.points.appendItem(p);
+    }
+}
+
+function debugRotations(svg, cx, cy, angle0, angle1) {
+    debugRotation(svg, cx, cy, angle0, "debug-rotation0");
+    debugRotation(svg, cx, cy, angle1, "debug-rotation1");
+}
+
+function translate(svg, svgElem, translation) {
+    getTransformsEnsuringFirstIsTranslation(svg, svgElem).getItem(0).setTranslate(translation.x, translation.y);
+}
+
+var rotations = {}
+function updateRotation(svg, svgElem, id, position0, position, angle0, angle1) {
+    debugRotations(svg, position.x, position.y, angle0, angle1);
+    var transforms = svgElem.transform.baseVal;
+    var rotate = null;
+    if (!(id in rotations)) {
+        rotate = svg.createSVGTransform();
+        svgElem.transform.baseVal.appendItem(rotate);
+        rotations[id] = rotate;
+    }
+    rotate = rotations[id];
+    // We append the rotation at the list of transforms,
```

Comment:
I wouldn't play with a list of transform, instead just create a transform matrix, it makes it simpler. Unfortunately there's no rotate with center method, but then you can do something like this:
```
rotate.setMatrix(svg.createSVGMatrix()
    .translate(translation)
    .translate(position0).rotate(rotation).translate(-position0));
```

And then for scaling you just need to insert a non-uniform scale at the right place.

Reply:
Yes, that will leave the operations we are performing explicit and open to further adjustments

Reply:
I've put all the required transformations in a single matrix as you suggested. Much more clear.
Just a side note: it seems at some point `SVGMatrix` will be replaced by `DOMMatrix` (however, it does not seem this will happen any time soon).